src/Autofac/Core/Resolving/Pipeline/ResolvePipelineBuilder.cs

```diff
@@ -270,34 +270,25 @@ Action<ResolveRequestContextBase> Chain(Action<ResolveRequestContextBase> next,
 
                 return (ctxt) =>
                 {
-                    // Optimise the path depending on whether diagnostics are enabled.
-                    if (ctxt.DiagnosticSource.MiddlewareDiagnosticsEnabled())
+                    ctxt.DiagnosticSource.MiddlewareStart(ctxt, stage);
```

Comment:
If there is a shortcut here for 'are any subscribers attached', using `IsEnabled` without the name parameter, can we use that to conditionally enter the entire diagnostic block vs the simple execution block, as per the original `MiddlewareDiagnosticsEnabled` check?  Just to save on the additional method calls we may not need to make.  

Unless you're thinking someone might add a diagnostic subscriber inside the middleware itself, and expect the Success/Failure events to then fire for that same middleware?

Reply:
Let me have one more look at this. It was pretty iterative to get where I got so I can see I missed that.

Reply:
I updated this to be closer to what we had to begin with:

- Hot path surrounded by `DiagnosticSource.IsEnabled()` (the null check) with one "version" of the pipeline being instrumented and the other being just the stage execution.
- Same thing in `ResolveOperationBase.GetOrCreateInstance` - but since there were more than a couple of lines I made the main pipeline invocation there an aggressively inlined method to avoid accidental copy/paste errors.
- Removed all the `IsEnabled()` checks from the extensions that write the diagnostics. Every write is now surrounded by a call to `IsEnabled()` and, where possible, that's only called one time around a block of code instead of at every potential write.

I can't guarantee it'll be _identical_ to the v6 perf but it should be... within tolerances? I can't imagine what else we could do to optimize the flow - it's boiling down to null checks like it was doing before.

Reply:
Would you like to run benchmarks again or should we just call it good?

Reply:
I'll run the benchmarks shortly and post back here, they only take a few moments to run.

Reply:
New Benchmarks. Still a ~0.5μs difference, but that might be fine.

**DeepGraphResolve - ~0.6μs / 6% slower**

|  Method |     Mean |     Error |    StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------- |---------:|----------:|----------:|-------:|-------:|------:|----------:|
| v6 | 9.539 us | 0.0693 us | 0.0615 us | 2.0599 | 0.0458 |     - |  16.92 KB |
| issue-788 | 10.18 us | 0.067 us | 0.063 us | 2.0599 | 0.0458 |     - |  16.91 KB |

**ChildScopeResolve - ~0.5μs / 2% slower**

|  Method |     Mean |    Error |   StdDev |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|-------- |---------:|---------:|---------:|-------:|-------:|------:|----------:|
| v6 | 31.43 us | 0.215 us | 0.202 us | 3.9063 | 1.9531 |     - |  32.09 KB |
| issue-788 | 32.09 us | 0.186 us | 0.165 us | 4.0894 | 2.0142 |     - |  33.88 KB |

Reply:
I'm not sure there's anything else I can do; it's a more robust diagnostics system than previous, and maybe that comes with a price. I'll commit it now and if we can figure out new ways to improve we can update then.

Reply:
Yeah, I don't think there's much more to do here, by virtue of a more robust system there's a few more method calls that have to take place, and that's that.

We can track any further investigation under #1127.